### ISSUE TITLE
print errors and messages to log file 

### ISSUE THREAD

**Issue by rnugent3** [HydrologicEngineeringCenter/HEC-FDA#710]
We need to design a way to print messages and errors to a log file that is helpful to the user, and reduces our technical debt. 

I think that all errors should be accessed through Validation, e.g. GetErrors(). The errors should not be passed as a message - that is a redundant way of getting to the errors. 

We should think about how to print the errors, and how to print the messages. What should the organization be? Errors, and then messages? Or errors and messages organized in a specific section for a given object?

**Comment by codymccoy** [comment]
I have been working on this. Here is what I have so far:
I think we should have a class that extends "Validation" that we call ValidationErrorLogger. Any class that we want to log errors, should extend this new class. That should be easy, because all those classes probably already extend Validation. I created this class in the Model, although I think it probably should go into the MVVM framework project.

![image](https://user-images.githubusercontent.com/47757848/210278028-b966f5a3-282a-4a5b-9e10-ca02c224bf7c.png)
Ignore the properties and the property rules, those are just for testing. 

![image](https://user-images.githubusercontent.com/47757848/210278063-52268ff6-6ced-4f32-87f2-65484592d37e.png)
I noticed that in several places you were adding an "intro" message before reporting the errors so I added a "LogErrors" that includes an intro message. This method gets the errors and reports them.

This is an example of how it can be used:
![image](https://user-images.githubusercontent.com/47757848/210279481-a3c764c0-a169-41b2-ab60-0fd246706837.png)

**Comment by codymccoy** [comment]
Issues with the above solution:
This doesn't work for any view model objects because those use their own validation system not the MVVM Framework. I think we tried to switch it all over to the new MVVM but that proved difficult. They were both written by Will and are very similar, it would just take some time to convert it all to one. We could duplicate the "ValidationErrorLogger" in the VM so that it works pretty much the same way, but then we are adding more parallel logic that would need to be unraveled down the road.

I don't see any good way of organizing the log file by type. I don't totally understand all the ins and outs of the message hub, but i think it just writes the messages as it gets them. We could probably have the message hub check some enum or what type the sender is (ie: ValidationErrorLogger) to determine if it is a message or a validation error. This way the log file could keep two different sections. But then you would have to do insertions to the log file at certain locations which i think probably involves rewriting the log file each time which we probably don't want to do.

**Comment by codymccoy** [comment]
I am also noticing double messages in the log file. I am not sure why.
https://github.com/HydrologicEngineeringCenter/HEC-FDA/issues/716

**Comment by codymccoy** [comment]
This work was done in branch Error_Validation

**Comment by rnugent3** [comment]
@HenryGeorgist do you have suggestions for how to organize messaging?

**Comment by rnugent3** [comment]
Sounds good, thank you. Do you have any thoughts on how we can write out the errors with some meta data about the object for which we are writing out the errors?

**Comment by rnugent3** [comment]
Like, a row with the object type and name, followed by the errors.

**Comment by rnugent3** [comment]
And ideally we'd do the same for messages

**Comment by rnugent3** [comment]
@HenryGeorgist what do you think?

**Comment by HenryGeorgist** [comment]
there is an error message in the mvvm framework.  https://github.com/HydrologicEngineeringCenter/Framework/blob/master/Model/Messaging/ErrorMessage.cs

there is also a debugging error message:
https://github.com/HydrologicEngineeringCenter/Framework/blob/master/Model/Messaging/DebuggingErrorMessage.cs

in the debugging error message i use runtime to get the caller member name (the method) line and file name which should mostly reflect the class name.


thoughts?

**Comment by codymccoy** [comment]
> Like, a row with the object type and name, followed by the errors.

The error logger could get the type, but I am not sure about the name. Is there an object or interface that defines the name property for your model objects?

**Comment by HenryGeorgist** [comment]
system.runtime.complierservices.membername and system.runtime.complerservices.filepath

**Comment by codymccoy** [comment]
I think @rnugent3 is talking about a user specified name. So not OccupancyType but RES1.

**Comment by codymccoy** [comment]
In the class I showed above, I added a string argument where you could pass in whatever you want. So you could inject the name of the object into there. For example "Errors with occtype RES1:" and then it would list the errors.

**Comment by rnugent3** [comment]
@codymccoy yes - a user specified name so the user knows where to look, and I like your example. do we need a study that has a bunch of errors to test?

**Comment by codymccoy** [comment]
Yes, a sample study would be nice.

**Comment by rnugent3** [comment]
@HenryGeorgist - I spoke with Cody about organizing error logging - his suggestion using ValidationErrorLogger in the screen shot above sounds like a good move to me - a step in the right direction. Cody opined (and I agree) that the class would go into Framework. Would you approve that PR on Framework?

**Comment by HenryGeorgist** [comment]
yep. is there a pr there already? just tag me. brennan might be an admin on that project too

**Comment by codymccoy** [comment]
No pr yet.

**Comment by codymccoy** [comment]
@rnugent3 I am running into an issue that I could use some help on. I changed some logic to use the new "logErrors()" that I wrote and it is throwing an exception.
Here is where i call log errors:
![image](https://user-images.githubusercontent.com/47757848/213570900-a433fdd0-fe0a-4ea1-857e-7da6b31ae80f.png)

In LogErrors I decided that we should probably validate one last time to see if there are any new errors. I figured it couldn't hurt but I was wrong. The act of validating is throwing an exception:
![image](https://user-images.githubusercontent.com/47757848/213571104-326ea469-df3e-4b4c-9f1f-09d0259144de.png)

It is evaluating an uncertainPairedData that has Xvals as null which is throwing a NPE.
![image](https://user-images.githubusercontent.com/47757848/213571208-bd28dc30-6c10-461a-8c80-933579fc9a6c.png)

I am not totally sure how that is possible, we should probably at least make those an empty array upon construction. The larger question i have is do we want to call validate in the log errors method or just log any existing errors?

**Comment by rnugent3** [comment]
@codymccoy would you please put 20-30min on the calendar to talk about this? PLease include @HenryGeorgist and @Brennan1994.

**Comment by HenryGeorgist** [comment]
we should eitther allocate an empty array, or check for nullness in the rule before checking length.

in the builder pattern for impactareascenario we had a build step, in the build step there should be a "validate" call on the simulation itself
https://github.com/HydrologicEngineeringCenter/HEC-FDA/blob/b28616e688540d279c999c05c3adb91909c7be0c/fda-model/compute/ImpactAreaScenarioSimulation.cs#L1119

if the appropriate rules get added when each element of the simulation get added, the validate method would validate and provide meaningful messages... example

https://github.com/HydrologicEngineeringCenter/HEC-FDA/blob/b28616e688540d279c999c05c3adb91909c7be0c/fda-model/compute/ImpactAreaScenarioSimulation.cs#L1167
